_posts/2020-05-07-removing-capybara-webkit.md

```diff
@@ -0,0 +1,133 @@
+-    field = find('#' + options[:from], visible: false)
+-    option_value = page.evaluate_script("$(\"##{field[:id]} option:contains('#{item_text}')\").val()")
+-    page.execute_script("value = ['#{option_value}']\; if ($('##{field[:id]}').val()) {$.merge(value, $('##{field[:id]}').val())}")
+-    option_value = page.evaluate_script("value")
+-    page.execute_script("$('##{field[:id]}').val(#{option_value})")
+-    page.execute_script("$('##{field[:id]}').trigger('liszt:updated').trigger('change')")
+
++    field = find_field(options[:from], visible: false)
++    find("##{field[:id]}_chzn").click
++    find("##{field[:id]}_chzn ul.chzn-results li", text: item_text).click
+  end
+```
+
+So we not only updated the gem, we also improved the test suite by not doing obscure Javascript calls.
+
+# Updating Docker Container
+
+Now that our tests are passing we can update the `Dockerfile` removing the commands that installed QT's libraries `qt5-default` and `libqt5webkit5-dev` (that will depend on the OS your are using, we have a Debian image so we just removed those packes from the `apt-get` command).
+
+Then we make sure we have Mozilla Firefox or Google Chrome installed. It was easier to install Firefox since the `firefox-esr` package was already available on the Debian Buster image we were using. You can use a Docker image which includes other browsers or install Google Chrome if you want to.
+
+So we changed the default driver to use Firefox:
+
+```ruby
+Capybara.javascript_driver = :headless_firefox
+```
+
+Now our tests can be run inside Docker and we are all green.
+
+# Updating CircleCI Config
+
+The last part of the process was to make sure the test passes on the CI system. We only needed to make sure Firefox was available when running the test, so the easiest solution for this was to use a CircleCI Ruby image and use the correct [variant](https://circleci.com/docs/2.0/circleci-images/#language-image-variants) so it also includes most used browsers:
+
+```diff
+- - image: circleci/ruby:2.4.10-buster-node
++ - image: circleci/ruby:2.4.10-buster-node-browsers
+```
+
+> We did have some problems related to caching, so you may want to try to retry your CI job without caching just to make sure if you see something failing
+
+## Conclusion
+
+We finally got an old dependency removed (QT) replacing `capybara-webkit` with a more modern solution.
+
+It makes it easier for new developers to join the project now that it has less extra requirements. It will also help when updating the Rails version, since newer Rails versions adds `webdrivers` gem when creating new apps by default.
+
+We also improved the test suite. When writing a feature spec (or system spec as introduced in Rails 5.2) you want to mimic the user interactions. If we don't test that the user can actually click an element and we trigger the `click` event using plain javascript, we may break something for the end user without noticing on the tests.
```

Comment:
break something for the end user?

Reply:
hmm I meant that if we are not testing the user interactions using capybara methods (and just modify the state using javascript) we could introduce some code that breaks things that a user should be able to interact with and our tests won't detect that

Reply:
maybe you just need to delete the word "end"?